### PR TITLE
[ALLUXIO-2236] File system worker thrift client pool

### PR DESCRIPTION
**Pull request by peisun1115**
https://alluxio.atlassian.net/browse/ALLUXIO-2236

### CODE REVIEW

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java [hunk 1]
```diff
@@ -227,48 +236,75 @@ public InetSocketAddress getWorkerDataServerAddress() {
...
+          }
+        });
   }
 
   /**
-   * Called only by {@link FileSystemWorkerClientHeartbeatExecutor}, encapsulates
-   * {@link #sessionHeartbeat()} in order to cancel and cleanup the heartbeating thread in case of
-   * failures.
+   * Sends a session heartbeat to the worker. This renews the client's lease on resources such as
+   * temporary files and updates the worker's metrics.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: Remove the metrics update part?

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java [hunk 2]
```diff
@@ -53,81 +56,81 @@
...
-  /** Underlying thrift RPC client which executes the operations. */
-  private FileSystemWorkerClientService.Client mClient;
-  /** The current heartbeat thread, this will be updated each time this client connects. */
-  private Future<?> mHeartbeat;
-  /** The current session id, managed by the caller. */
-  private long mSessionId;
+  private ScheduledFuture<?> mHeartbeat = null;
+
+  private static final ConcurrentHashMapV8<InetSocketAddress, FileSystemWorkerThriftClientPool>
+      CLIENT_POOL = new ConcurrentHashMapV8<>();
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: These should be `POOLS`?

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java [hunk 3]
```diff
@@ -227,48 +236,75 @@ public InetSocketAddress getWorkerDataServerAddress() {
...
+   * Sends a session heartbeat to the worker. This renews the client's lease on resources such as
+   * temporary files and updates the worker's metrics.
+   *
+   * @throws IOException if an I/O error occurs
    */
-  public synchronized void periodicHeartbeat() {
-    if (mClosed) {
-      return;
-    }
+  public void sessionHeartbeat() throws IOException {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: Should this throw an interrupted exception?

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java (hunk 2 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: done.

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java (hunk 3 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: done

**Comment by peisun1115** [comment]
@calvinjia Thanks for the review. I have addressed the comments. PTAL.

**Comment by calvinjia** [comment]
@peisun1115 looks good overall. It may make sense to make specific `acquireInternal` calls so the interrupted exception only needs to be propagated for some of them?

**Comment by peisun1115** [comment]
@calvinjia I've added another one.

**Pull request merged**